mesonbuild/interpreter.py

```diff
@@ -151,6 +156,126 @@ def prepend_method(self, args, kwargs):
         self.add_var(self.held_object.prepend, args, kwargs)
 
 
+class Component(MutableInterpreterObject):
+    def __init__(self, name, interpreter, option_name, dependencies=None, silent=False):
+        super().__init__()
+
+        self.name = name
+        self.interpreter = interpreter
+        self.option_name = option_name
+        self.targets = set()
+        if dependencies is None:
+            self.dependencies = set()
+        else:
+            self.dependencies = set(dependencies)
+
+        self.methods.update({'targets': self.func_targets,
+                             'enabled': self.func_enabled,
+                             'value': self.func_value,
+                             'add_dependencies': self.func_add_dependency,
+                             'dependencies': self.func_dependencies})
+
+        value = self.interpreter.environment.coredata.user_options[self.option_name].value
+        if value == 'enabled':
+            missing_deps = self.check_dependencies()[1]
+            if missing_deps:
+                raise InterpreterException("Component %s is enabled but the following"
+                                           " dependencies are not present: %s" % (
+                                               name, missing_deps))
+            value = mlog.green(value)
+        elif self.func_enabled():
+            value = mlog.green(value)
+        else:
+            value = mlog.red(value)
+
+        if not silent:
+            mlog.log('Component', mlog.bold(name), ':', value)
+
+    def func_enabled(self, *args, **kwargs):
+        return self.check_dependencies()[0]
+
+    def func_value(self, *args, **kwargs):
+        return self.interpreter.environment.coredata.user_options[self.option_name].value
+
+    def func_add_dependency(self, args, kwargs):
+        for dep in mesonlib.flatten(args):
+            self.add_dependency(dep)
+
+    def add_dependency(self, dep):
+        dep = getattr(dep, "held_object", dep)
+        if not isinstance(dep, dependencies.Dependency):
+            raise InterpreterException("component.add_dependencies only takes "
+                                       " dependency objects as argument (not: %s)" % (
+                                           dep))
+        value = self.interpreter.environment.coredata.user_options[self.option_name].value
+        if value == 'enabled' and not dep.found():
+            raise InterpreterException("Component %s is enabled but the following"
+                                       " dependencies are not present: %s" % (
+                                           self.name, dep.get_name()))
+
+        if dep not in self.dependencies:
+            if not dep.found() and self.func_enabled():
+                mlog.log('component', mlog.bold(self.name), 'is now disabled by:',
+                        mlog.red(dep.name))
```

Comment:
__[flake8]__
* [E128] continuation line under-indented for visual indent

<!-- Comment by [SideCI](https://sideci.com) -->